File: code/smv-debugger/src/main/java/org/fbme/smvDebugger/panel/mvc/StepIndexModel.java
package org.fbme.smvDebugger.panel.mvc;

import javax.swing.event.SwingPropertyChangeSupport;
import java.beans.PropertyChangeListener;

public class StepIndexModel {
  public static final String STEP_INDEX = "stepIndex";

  private final SwingPropertyChangeSupport propertyChangeSupport = new SwingPropertyChangeSupport(this);
  private int currentStepIndex;
  private int previousStepIndex;

  public int getCurrentStepIndex() {
    return currentStepIndex;
  }

  public int getPreviousStepIndex() {
    return previousStepIndex;
  }

  public void setStepIndex(final int stepIndex) {
    previousStepIndex = currentStepIndex;
    currentStepIndex = stepIndex;
    propertyChangeSupport.firePropertyChange(STEP_INDEX, previousStepIndex, currentStepIndex);
  }

  public void addPropertyChangeListener(final PropertyChangeListener listener) {
    propertyChangeSupport.addPropertyChangeListener(listener);
  }

  public void addPropertyChangeListener(final String propertyName, final PropertyChangeListener listener) {
    propertyChangeSupport.addPropertyChangeListener(propertyName, listener);
  }
}

File: code/smv-debugger/src/main/kotlin/org/fbme/smvDebugger/panel/mvc/StepIndexModel.kt
package org.fbme.smvDebugger.panel.mvc

import java.beans.PropertyChangeListener
import javax.swing.event.SwingPropertyChangeSupport

class StepIndexModel {
    private val propertyChangeSupport = SwingPropertyChangeSupport(this)
    var currentStepIndex = 0
        private set
    var previousStepIndex = 0
        private set

    fun setStepIndex(stepIndex: Int) {
        previousStepIndex = currentStepIndex
        currentStepIndex = stepIndex
        propertyChangeSupport.firePropertyChange(STEP_INDEX, previousStepIndex, currentStepIndex)
    }

    fun addPropertyChangeListener(listener: PropertyChangeListener?) {
        propertyChangeSupport.addPropertyChangeListener(listener)
    }

    fun addPropertyChangeListener(propertyName: String?, listener: PropertyChangeListener?) {
        propertyChangeSupport.addPropertyChangeListener(propertyName, listener)
    }

    companion object {
        const val STEP_INDEX = "stepIndex"
    }
}
